src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IsPatternOperator.cs

```diff
@@ -24,10 +24,12 @@ public override BoundNode VisitIsPatternExpression(BoundIsPatternExpression node
                 result = isPatternRewriter.LowerIsPatternAsLinearTestSequence(node, decisionDag, whenTrueLabel: node.WhenTrueLabel, whenFalseLabel: node.WhenFalseLabel);
                 isPatternRewriter.Free();
             }
-            else if (canProduceLinearSequence(decisionDag.RootNode, whenTrueLabel: node.WhenFalseLabel, whenFalseLabel: node.WhenTrueLabel))
+            else if (IsFailureNode(decisionDag.RootNode, node.WhenFalseLabel))
             {
-                // If we can build a linear test sequence with the whenTrue and whenFalse labels swapped, then negate the
-                // result.  This would typically arise when the source contains `e is not pattern`.
+                // If the given pattern always fails due to a constant input (see comments on BoundDecisionDag.SimplifyDecisionDagIfConstantInput),
+                // we build a linear test sequence with the whenTrue and whenFalse labels swapped and then negate the result, to keep the result a constant.
+                // Note that the positive case will be handled by canProduceLinearSequence above, however, we avoid to produce a full inverted linear sequence here
+                // because we may be able to generate better code for a sequence of `or` patterns, using a switch dispatch, for example, which is done in the general rewriter.
                 negated = !negated;
```

Comment:
>negated = !negated; [](http://example.com/codeflow?start=16&length=19)

Is there a targeted test covering this code path now?

Reply:
That would be `IsNot_18`. Let me know if there's any other scenario you want to verify. Thanks.

Reply:
Took me a while to understand what you were concerning about.
- In the first `if` we check if there's a single path to the true label with all the other paths leading to false label.
- In the second `if` we check if there's a *direct* path to the false label (in which case there would be no true label).

We still need to negate the result here because whenTrue and whenFalse are swapped for the linear rewriter as it expects whenTrue to be the "final" leaf node.